Add Postgres integration test that exercises deadlocks

MSch opened the pull request:
[I noticed](https://github.com/MSch/ecto/blob/2b8ba379a5ff29f6b4847955d5d9c3372935b897/integration_test/pg/deadlock_test.exs#L50-L54) that Ecto's worker don't notice if Postgres aborts the transaction and thought this would lead to bugs down the road, but after investigating some more (and writing a test case) I think there are no problems (because Postgres silently accepts COMMITs for aborted transactions)

ericmj commented:
This is great @MSch.

One thing to keep in mind with Postgres transactions is that when a transaction is aborted it is completely aborted until fully commited or fully rollbacked, regardless of savepoints. This does not map perfectly to ecto's transactions since we model nested transactions with savepoints. This may not be an issue in practice but could be useful to keep in mind.

ericmj commented:
Should we add a test with nested transactions?

MSch commented:
@ericmj I tried to add a test case for nested transactions but couldn't really figure out what was going on:

```
18:09:21.742 [error] GenServer #PID<0.164.0> terminating
** (exit) an exception was raised:
    ** (ErlangError) erlang error: :"function not exported"
        (elixir) GenServer.reply({:error, %Postgrex.Error{message: nil, postgres: %{code: "25P02", file: "postgres.c", line: "1256", message: "current transaction is aborted, commands ignored until end of transaction block", routine: "exec_parse_message", severity: "ERROR"}}})
        (ecto) lib/ecto/adapters/sql/worker.ex:125: Ecto.Adapters.SQL.Worker.handle_call/3
        (stdlib) gen_server.erl:607: :gen_server.try_handle_call/4
        (stdlib) gen_server.erl:639: :gen_server.handle_msg/5
        (stdlib) proc_lib.erl:237: :proc_lib.init_p_do_apply/3

18:09:21.744 [error] Task #PID<0.179.0> started from #PID<0.178.0> terminating
Function: #Function<1.35622615/0 in Ecto.Integration.DeadlockTest.test deadlocks reset worker/1>
    Args: []
** (exit) exited in: GenServer.call(#PID<0.164.0>, :unlink_me, 5000)
    ** (EXIT) no process
```

I think it is related to `wipe_state` somehow.

MSch commented:
Also, with savepoints Postgres' deadlock detector does not abort the transaction but only the "nested' transaction, rolling back to the savepoint makes the tx viable again (and it has kept its locks)

ericmj commented:
@MSch That's not what I found from my tests a while ago, but I could misremember or made the tests wrong.

josevalim commented:
@MSch The error is related to the best GenServer.reply, your PR has been merged. Can you take another look now?

MSch commented:
@ericmj I tried it again before my comment, but I'll write a test for that case too to make sure and to not forget again.

@josevalim yeah, I noticed that I fixed that when investigating nested txs but forgot that I removed the commit from this PR

MSch commented:
Here's what I'm getting after rebasing on master:

```
18:17:32.776 [error] GenServer #PID<0.605.0> terminating
** (exit) an exception was raised:
    ** (FunctionClauseError) no function clause matching in Ecto.Adapters.SQL.Worker.handle_call/3
        (ecto) lib/ecto/adapters/sql/worker.ex:84: Ecto.Adapters.SQL.Worker.handle_call({:rollback, [timeout: 5000]}, {#PID<0.618.0>, #Reference<0.0.0.10932>}, %{conn: #PID<0.622.0>, link: nil, module: Ecto.Adapters.Postgres.Connection, params: [otp_app: :ecto, username: "postgres", password: "postgres", database: "ecto_test", hostname: "localhost"], sandbox: false, transactions: 0})
        (stdlib) gen_server.erl:607: :gen_server.try_handle_call/4
        (stdlib) gen_server.erl:639: :gen_server.handle_msg/5
        (stdlib) proc_lib.erl:237: :proc_lib.init_p_do_apply/3


  1) test deadlocks reset worker (Ecto.Integration.DeadlockTest)
     integration_test/pg/deadlock_test.exs:9
     ** (exit) exited in: GenServer.call(#PID<0.605.0>, :unlink_me, 5000)
         ** (EXIT) no process
     stacktrace:
       (elixir) lib/gen_server.ex:356: GenServer.call/3
       (ecto) lib/ecto/adapters/sql.ex:460: Ecto.Adapters.SQL.checkin_worker/2
       (ecto) lib/ecto/adapters/sql.ex:422: Ecto.Adapters.SQL.transaction/3
       integration_test/pg/deadlock_test.exs:19
```

MSch commented:
@ericmj here's the test case demonstrating that if a nested tx aborts and rolls back the parent tx keeps its locks: https://github.com/MSch/ecto/blob/nested-tx-locking-test/integration_test/pg/deadlock_test.exs#L30 I'm not sure if this actually belongs in Ecto though.

josevalim commented:
@ericmj I am a bit lost in here, can I leave this issue assigned to you and you will merge it when you find appropriate?

MSch commented:
@ericmj I've removed the (failing) test for nested txs, I think this PR can be merged already even without testing nested txs and I'll open one for nested txs.

ericmj commented:
Merged #444.
